### ISSUE TITLE
Bug: [no-unecessary-type-assertion] false positive when types used for inference

### ISSUE THREAD

**Issue by mrginglymus** [typescript-eslint/typescript-eslint#12250]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.tsx&code=C4TwDgpgBAKhDOwA8MoQB7AgOwCbygCUIBjAewCdclEKBLbAcwBooBXbAa2zIHdsAfAKgBeKAG8AvgFgAUHIBmHEsDplsUBgogUKEaqgxY8BOIiTYIANx0CA2gF0BACgD6DMG2AAuCVACGuLjq8AD8vqiSAJS%2B4nJQUGRgqiERaJg4%2BLAIyFo6UADC9g5QoYVQvpY2FHKSEvFQesBsFBpxsgkJSSnY8LF1-qbpxllmudjaFIXFpeWV1joNMrLLcuS9wFBYiKKaEzp6uM7iAUGpUHZSAaY5SCdg-hT%2BALYQWBR9fmBkwP7AZL4AEZkMgAGwg-g0dUkAhK0QA3HI1iEwRAAHSgsiMZzbYBo7pqXpoh5PV7veDEn5-MhRJHyWTrHa4gBMuzyun0x1OwV6vkukjhUURDJR4IxWJxOWZ%2BOShIpJJebx08qp-yiQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0YANrhsOOdA7RIGzQF0NAXxAmgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
type Test<T extends Record<string, unknown>> = {}

function inferred<T extends Test<never>[]>(_input: { addons?: T }): {
  options: T extends Test<infer C>[] ? C : never
} {
  return {
    options: {} as T extends Test<infer C>[] ? C : never
  }
}

const test = inferred({ addons: [{} as Test<{ parameters: { potato: boolean } }>] });

console.log(test.options.parameters.potato)


const test2 = inferred({ addons: [{}] });
console.log(test2.options.parameters.potato)
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-type-assertion": ["error", ...<options>],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

assertion is **not** marked as unnecessary

### Actual Result

assertion is marked as unnecessary

### Additional Info

`definePreview` infers its return type from the `PreviewAddon` types passed to `addons`.

typescript eslint correctly identifies that this casting is not necessary from the perspective of assignment, but it _is_ required from the perspective of inference.

**Comment by bradzacher** [comment]
Could you please create an _**isolated**_ reproduction? As in -- without importing 3rd party types?

The playground doesn't load properly or show the errors you want.
In fact for whatever reason it looks like the playground trying to import those 3rd party modules causes the browser (tried in FF and chrome) to lock up entirely...

Also -- we can't add test cases importing 3rd party modules.

**Comment by mrginglymus** [comment]
I'll do so as soon as I can, but I wanted to highlight (along with the other three issues) that the changes to this rule may need reconsidering given the breadth of issues they've generated.

**Comment by bradzacher** [comment]
> breadth of issues they've generated

It's generated 2 _confirmed_ bugs, one crash (which has been fixed), and this yet to be confirmed issue.
That's hardly much in the way of "breadth"?

With any change to a highly used type-aware rule there is going to be issues -- it's entirely expected as the scope of what TS types allow you to express is so wide and varied that it's physically impossible for us to cover everything with tests.

The implementing PR added 75 test cases (1000 lines of tests) -- increasing the test case count by 20% (and line count by 64%). The rule itself has a test coverage of 95% test coverage. All of that and still there's some gaps.

Ultimately we can't just keep ideating about all the different cases that might potentially have different behaviour in the rule -- we need to release it and rely on users to provide us with missed cases.

**Comment by mrginglymus** [comment]
https://typescript-eslint.io/play/#ts=5.5.2&fileType=.tsx&code=C4TwDgpgBAKhDOwA8MoQB7AgOwCbygCUIBjAewCdclEKBLbAcwBooBXbAa2zIHdsAfAKgBeKAG8AvgFgAUHIBmHEsDplsUBgogUKEaqgxY8BOIiTYIANx0CA2gF0BACgD6DMG2AAuCVACGuLjq8AD8vqiSAJS%2B4nJQUGRgqiERaJg4%2BLAIyFo6UADC9g5QoYVQvpY2FHKSEvFQesBsFBpxsgkJSSnY8LF1-qbpxllmudjaFIXFpeWV1joNMrLLcuS9wFBYiKKaEzp6uM7iAUGpUHZSAaY5SCdg-hT%2BALYQWBR9fmBkwP7AZL4AEZkMgAGwg-g0dUkAhK0QA3HI1iEwRAAHSgsiMZzbYBo7pqXpoh5PV7veDEn5-MhRJHyWTrHa4gBMuzyun0x1OwV6vkukjhUURDJR4IxWJxOWZ%2BOShIpJJebx08qp-yiQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0YANrhsOOdA7RIGzQF0NAXxAmgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

Here's a slightly more isolated example.

Apologies for my brusqueness - I really do appreciate the work that goes into this plugin, which is why I do try to keep up to date and raise issues where I find them. And I didn't at all want to suggest that that the implementing PR was at all lacking in quality - bugs will get through.

I raise these somewhat incomplete issues partially for the benefit of the maintianers, but also for the benefit of other users who are updating. Finding that someone else has had the same problem as you is invaluable.

**Comment by GeKorm** [comment]
Even smaller example if it helps

```ts
type Override = (value: number) => void;

const foo = async (value: number) => {
  await fetch(`/${value}`);
}

[1, 2, 3].forEach(foo as Override);
```

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.tsx&code=C4TwDgpgBA8gbhAToglgE2gXigCjgQwBsBXCALigDtiBbAIyQEopMA%2BKOAe3QG4BYAFCCAxp0oBnYFABmnTiyj5xISsNwES5KrQaJmbKAG9BURQHd8KKdIjBhACxwADAPQASQxtIBfJ434C3oKCANoAjAA0UABMUQDMALoAdLKIAKL4Djiy8kqwCMjoEP6CQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0YANrhsOOdA7RIGzQF0ANAagES5KjXrJGLdlwC2lZLFkATWsR0vZqDHVNLWgdPTNDDQBfECigA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

Still an error, but the receiver does not accept the original type of the expression as indicated by `no-unnecessary-type-assertion`.

**Comment by bradzacher** [comment]
@GeKorm you can delete the type assertion from your example and the code typechecks fine without any loss of safety or correctness.

Thus the rule is right - the annotation is unnecessary.

The fact that removing the annotation causes a `no-misused-promises` error is also good and is also fully correct - you were previously hackily hiding this linting error using an assertion to trick that rule.

---

Your example is also entirely unrelated to OP, which is specifically about an edge case use of types to propagate generics through a system.

**Comment by bradzacher** [comment]
@mrginglymus this is an interesting case. For all intents and purposes your annotation is unnecessary - the type that TS eventually resolves to is in fact `{}` - meaning you are writing the code `{} as {}`.

The thing that's lost here is that TS does do something magic - it acts lazily. By being lazy and refusing to resolve the type alias to its underlying type till the last moment it incidentally also allows it to propagate the generic to the consumers who can read and operate on the type.

In other languages this is generally referred to as a "phantom" type - a type that isn't really there but also kind of is. It's an interesting thing!

This pattern is something very rarely used because the code is usually smelly - requiring `as` assertions and/or unused generics parameters.

**Comment by mrginglymus** [comment]
Yeah, it's definitely a non-ideal solution for this situation...and I will probably take this as a hint to switch to a more robust approach.
